bakery/checker.go

```diff
@@ -397,98 +345,81 @@ func (a *AuthChecker) allowAny(ctx context.Context, ops []Op) (authed, used []bo
-func (a *allowContext) checkIndirect(ctx context.Context) {
+func (a *allowContext) checkIndirect(ctx context.Context) []checkers.Caveat {
 	if a.checker.p.OpsAuthorizer == nil {
-		return
+		return nil
 	}
+	var allCaveats []checkers.Caveat
 	for op, mindexes := range a.checker.authIndexes {
 		if len(a.need) == 0 {
-			return
+			break
 		}
-		authedOK, err := a.checker.p.OpsAuthorizer.AuthorizeOps(ctx, op, a.need)
-		if err != nil {
-			// TODO this probably means "can't check" rather than "authorization denied";
-			// perhaps we should return rather than carrying on?
-			a.addError(err)
-			continue
-		}
-		for i, ok := range authedOK {
-			if !ok {
+		for _, mindex := range mindexes {
+			if a.status[mindex]&statusOK == 0 {
 				continue
 			}
-			// This operation is potentially authorized. See whether we have a macaroon
-			// that actually allows this operation.
-			for _, mindex := range mindexes {
-				if _, err := a.checker.checkConditions(ctx, a.need[i], a.checker.conditions[mindex]); err != nil {
-					a.addError(err)
+			ctx := checkers.ContextWithMacaroons(ctx, a.checker.Namespace(), a.checker.macaroons[mindex])
+			authedOK, caveats, err := a.checker.p.OpsAuthorizer.AuthorizeOps(ctx, op, a.need)
+			if err != nil {
+				// TODO this probably means "can't check" rather than "authorization denied";
+				// perhaps we should return rather than carrying on?
+				a.addError(err)
+				continue
+			}
+			// TODO we could perhaps combine identical third party caveats here.
+			allCaveats = append(allCaveats, caveats...)
+			for i, ok := range authedOK {
+				if !ok {
 					continue
 				}
 				// Operation is authorized. Mark the appropriate macaroon as used,
 				// and remove the operation from the needed list so that we don't
 				// bother AuthorizeOps with it again.
-				a.used[mindex] = true
+				a.status[mindex] |= statusUsed
 				a.authed[a.needIndex[i]] = true
+				a.opIndexes[a.need[i]] = mindex
 			}
 		}
 		a.updateNeed()
 	}
-}
-
-// checkWithAuthorizer checks which operations are authorized by the
-// Authorizer instance. We call Authorize even when we haven't got an
-// authenticated identity.
-func (a *allowContext) checkWithAuthorizer(ctx context.Context) ([]checkers.Caveat, error) {
-	oks, caveats, err := a.checker.p.Authorizer.Authorize(ctx, a.checker.identity, a.need)
+	if len(a.need) == 0 {
+		return allCaveats
+	}
+	// We've still got at least one operation unauthorized.
+	// Try to see if it can be authorized with no operation at all.
+	authedOK, caveats, err := a.checker.p.OpsAuthorizer.AuthorizeOps(ctx, NoOp, a.need)
 	if err != nil {
-		// TODO if there are macaroons supplied that have failed, perhaps we shouldn't
-		// do this but return those errors instead? Doing things the current
-		// way means that we lose the previous errors.
-		return nil, errgo.Notef(err, "cannot check permissions")
+		// TODO this probably means "can't check" rather than "authorization denied";
```

Comment:
i'd return internal server error in this case....  i'd definitely want to monitor this, though..

Reply:
changed to return an error (and updated the AuthorizeOps comment too)